File: solutions/java/bob/2/src/main/java/Bob.java
class Bob {

    String hey(String input) {

        input = input.trim();

        if (input.isEmpty()) {
            return "Fine. Be that way!";
        }

        if (isQuestion(input) && isShouting(input)) {
            return "Calm down, I know what I'm doing!";
        }

        if (isQuestion(input)) {
            return "Sure.";
        }

        if (isShouting(input)) {
            return "Whoa, chill out!";
        }

        return "Whatever.";
    }

    private boolean isQuestion(String input) {
        return input.endsWith("?");
    }

    private boolean isShouting(String input) {
        return input.toUpperCase().equals(input) && input.chars().anyMatch(Character::isLetter);
    }

}

File: solutions/haskell/bob/2/src/Bob.hs
{-# LANGUAGE OverloadedStrings #-}
module Bob (responseFor) where

import qualified Data.Char as C
import qualified Data.Text as T

import Data.Text (Text)

responseFor :: Text -> Text
responseFor text
  | isSilence          = "Fine. Be that way!"
  | isShoutingQuestion = "Calm down, I know what I'm doing!"
  | isShouting         = "Whoa, chill out!"
  | isQuestion         = "Sure."
  | otherwise          = "Whatever."
  where
    isShouting         = T.any C.isAlpha text && T.toUpper text == text
    isQuestion         = T.isSuffixOf "?" (T.stripEnd text)
    isShoutingQuestion = isShouting && isQuestion
    isSilence          = T.null (T.strip text)
